File: Lab3/v4_final/Python/projectile.py
# projectile.py - Снаряды игрока и врагов

import pygame
import math
from config import *

class Projectile:
    def __init__(self, x, y, angle, speed=PROJECTILE_SPEED, 
                 color=PROJECTILE_COLOR_PLAYER, is_player_shot=True):
        self.x = x
        self.y = y
        self.angle = angle
        self.speed = speed
        self.color = color
        self.lifetime = PROJECTILE_LIFETIME
        self.radius = PROJECTILE_RADIUS
        self.is_player_shot = is_player_shot
    
    def update(self):
        """Обновление позиции снаряда"""
        self.x += math.cos(self.angle) * self.speed
        self.y += math.sin(self.angle) * self.speed
        self.lifetime -= 1
    
    def collides_with(self, obstacle):
        """Проверка столкновения с объектом"""
        # Берега и острова
        if hasattr(obstacle, 'contains_point') and obstacle.contains_point(self.x, self.y, self.radius):
            return True
        
        # Враги и другие объекты с координатами и радиусом
        if hasattr(obstacle, 'x') and hasattr(obstacle, 'y') and hasattr(obstacle, 'radius'):
            dx = self.x - obstacle.x
            dy = self.y - obstacle.y
            distance = math.sqrt(dx*dx + dy*dy)
            return distance < self.radius + obstacle.radius
        
        return False
    
    def draw(self, screen, camera_y):
        """Отрисовка снаряда"""
        y_screen = int(self.y - camera_y)
        pygame.draw.circle(screen, self.color, (int(self.x), y_screen), self.radius)

File: Lab4_1/flutter_joystick_app/lib/game/models/projectile.dart
import 'dart:math';
import '../game_config.dart';
import 'island.dart';
import 'player.dart';
import 'enemy.dart';

class Projectile {
  double x;
  double y;
  double angle;
  late double speed;
  late int color;
  late int lifetime;
  late double radius;
  bool isPlayerShot;

  Projectile(this.x, this.y, this.angle, this.isPlayerShot,
      {double? speed, int? color}) {
    this.speed = speed ?? GameConfig.projectileSpeed;
    this.color = color ?? (isPlayerShot 
        ? GameConfig.projectileColorPlayer 
        : GameConfig.projectileColorEnemy);
    lifetime = GameConfig.projectileLifetime;
    radius = GameConfig.projectileRadius;
  }

  void update() {
    x += cos(angle) * speed;
    y += sin(angle) * speed;
    lifetime--;
  }

  bool collidesWith(dynamic obstacle) {
    if (obstacle is Island) {
      return obstacle.collidesWith(x, y, radius);
    } else if (obstacle is Shore) {
      return obstacle.collidesWith(x, y, radius);
    } else if (obstacle is Enemy) {
      double dx = x - obstacle.x;
      double dy = y - obstacle.y;
      double distance = sqrt(dx * dx + dy * dy);
      return distance < radius + obstacle.radius;
    } else if (obstacle is Player) {
      double dx = x - obstacle.x;
      double dy = y - obstacle.y;
      double distance = sqrt(dx * dx + dy * dy);
      return distance < radius + obstacle.radius;
    }
    return false;
  }
}

